tests/testping.nim

```diff
@@ -0,0 +1,140 @@
+import options, bearssl
+import chronos, strutils
+import ../libp2p/[protocols/identify,
+                  protocols/ping,
+                  multiaddress,
+                  peerinfo,
+                  wire,
+                  peerid,
+                  stream/connection,
+                  multistream,
+                  transports/transport,
+                  transports/tcptransport,
+                  crypto/crypto,
+                  upgrademngrs/upgrade]
+import ./helpers
+
+when defined(nimHasUsed): {.used.}
+
+suite "Ping":
+  teardown:
+    checkTrackers()
+
+  suite "handle ping message":
+    var
+      ma {.threadvar.}: MultiAddress
+      serverFut {.threadvar.}: Future[void]
+      acceptFut {.threadvar.}: Future[void]
+      pingProto1 {.threadvar.}: Ping
+      pingProto2 {.threadvar.}: Ping
+      transport1 {.threadvar.}: Transport
+      transport2 {.threadvar.}: Transport
+      msListen {.threadvar.}: MultistreamSelect
+      msDial {.threadvar.}: MultistreamSelect
+      conn {.threadvar.}: Connection
+      pingReceivedCount {.threadvar.}: int
+
+    asyncSetup:
+      ma = Multiaddress.init("/ip4/0.0.0.0/tcp/0").tryGet()
+
+      transport1 = TcpTransport.init(upgrade = Upgrade())
+      transport2 = TcpTransport.init(upgrade = Upgrade())
+
+      proc handlePing(peer: PeerInfo) {.async, gcsafe, closure.} =
+        inc pingReceivedCount
+      pingProto1 = newPing()
+      pingProto2 = newPing(handlePing)
+
+      msListen = newMultistream()
+      msDial = newMultistream()
+
+      pingReceivedCount = 0
+
+    asyncTeardown:
+      await conn.close()
+      await acceptFut
+      await transport1.stop()
+      await serverFut
+      await transport2.stop()
+
+    asyncTest "simple ping":
+      msListen.addHandler(PingCodec, pingProto1)
+      serverFut = transport1.start(ma)
+      proc acceptHandler(): Future[void] {.async, gcsafe.} =
+        let c = await transport1.accept()
+        await msListen.handle(c)
+
+      acceptFut = acceptHandler()
+      conn = await transport2.dial(transport1.ma)
+
+      discard await msDial.select(conn, PingCodec)
+      let time = await pingProto2.ping(conn)
+
+      check not time.isZero()
+
+    asyncTest "networked cancel ping":
+      let baseMa = Multiaddress.init("/ip4/127.0.0.1/tcp/0").tryGet()
+
+      let transport: TcpTransport = TcpTransport.init(upgrade = Upgrade())
+      let transportdialer: TcpTransport = TcpTransport.init(upgrade = Upgrade())
+      asyncCheck transport.start(baseMa)
+
+      proc acceptHandler() {.async, gcsafe.} =
+        let conn = await transport.accept()
+        #await conn.write("Hello dns!")
+        var buf: array[32, byte]
+        await conn.readExactly(addr buf[0], 32)
+        await conn.write(addr buf[0], 32)
+
+      let handlerWait = acceptHandler()
+
+      let streamTransport = await transportdialer.dial(transport.ma)
+
+      let p = pingProto2.ping(streamTransport)
```

Comment:
This is actually pretty tricky - as it is, this test isn't really testing cancellation there are no guarantees that by the time you call `cancelAndWait` the sequence hasn't already completed. What needs to happen is that right after accepting the connection we await an "empty" future and then call `p.cancelAndAwait` at least then, we know for sure that we're cancelling in the middle of the sequence and even then we're only properly testing the caller.

Unfortunately we can't test all the cancellation points, specially on the receiver end and we potentially need something like #520 to do proper fuzzing.

Reply:
My understanding is that the `cancelAndWait` will be called just after the first `await` of `ping` (so just after it sent the data).
Since chronos is single threaded, this should be always the case. To test at each step we could do something like:
```nim
for pollCount in 0..20:
   setupPingTest()
   let future = pingProto2.ping(streamTransport)
   for _ in 0..<pollCount:
      if future.finished: break
      poll()
   if future.finished: break #We actually finished the sequence
   await future.cancelAndWait()
```

This way, we're `poll`'ing one more time at each loop, and thus testing the cancellation in a further point of the sequence each time. But I haven't dug into chronos, so maybe I missed something

Reply:
The idea is that we at least test it at a point where it's actually done something. Cancelling right after the ping will potentially happen before any real work has begun. Cancelling after the connection has been established will have more cleanup to do and thus will end up testing more of the flow.

I'm curious to see what the snippet actually does. It doesn't seem it will get past a specific point it the cancelation?